docs/ansible/coding-conventions.md

```diff
@@ -0,0 +1,669 @@
+      lookup_plugins/     # or other types of plugins, like lookup in this case
+
+    webtier/              # same kind of structure as "common" was above, done for the webtier role
+    monitoring/           # ""
+    fooapp/               # ""
+```
+
+## Alternative Directory Layout
+
+Alternatively you can put each inventory file with its group_vars/host_vars in a separate directory. This is particularly useful if your group_vars/host_vars don’t have that much in common in different environments. The layout could look something like this:
+
+```text
+inventories/
+   production/
+      hosts               # inventory file for production servers
+      group_vars/
+         group1.yml       # here we assign variables to particular groups
+         group2.yml
+      host_vars/
+         hostname1.yml    # here we assign variables to particular systems
+         hostname2.yml
+
+   staging/
+      hosts               # inventory file for staging environment
+      group_vars/
+         group1.yml       # here we assign variables to particular groups
+         group2.yml
+      host_vars/
+         stagehost1.yml   # here we assign variables to particular systems
+         stagehost2.yml
+
+library/
+module_utils/
+filter_plugins/
+
+site.yml
+webservers.yml
+dbservers.yml
+
+roles/
+    common/
+    webtier/
+    monitoring/
+    fooapp/
+```
+
+This layout gives you more flexibility for larger environments, as well as a total separation of inventory variables between different environments. The downside is that it is harder to maintain, because there are more files.
+
+## Group And Host Variables
+
+Groups are nice for organization, but that’s not all groups are good for.
+You can also assign variables to them! For instance, atlanta has its own NTP servers, so when setting up ntp.conf, we should use them.
+Let’s set those now:
+
+```yaml
+---
+# file: group_vars/atlanta
+ntp: ntp-atlanta.example.com
+backup: backup-atlanta.example.com
+```
+
+Variables aren’t just for geographic information either! Maybe the webservers have some configuration that doesn’t make sense for the database servers:
+
+```yaml
+---
+# file: group_vars/webservers
+apacheMaxRequestsPerChild: 3000
+apacheMaxClients: 900
+```
+
+If we had any default values, or values that were universally true, we would put them in a file called group_vars/all:
+
+```yaml
+---
+# file: group_vars/all
+ntp: ntp-boston.example.com
+backup: backup-boston.example.com
+```
+
+We can define specific hardware variance in systems in a host_vars file, but avoid doing this unless you need to:
+
+```yaml
+---
+# file: host_vars/db-bos-1.example.com
+foo_agent_port: 86
+bar_agent_port: 99
+```
+
+Again, if we are using dynamic inventory sources, many dynamic groups are automatically created.
+So a tag like “class:webserver” would load in variables from the file “group_vars/ec2_tag_class_webserver” automatically.
+
+## Top Level Playbooks Are Separated By Role
```

Comment:
didn't get the idea of this section at all. we usually do not write such things. let's remove for now, will return in case we will find out that we need to include playbooks